Added OpportunisticResourceEstimator module

treefell opened the pull request:
Added the creation of **opportunisticResourceEstimator** inheriting **resourceEstimator** module.
Using the **CommandRunner** from _Mesos-Command-Module_ to call a binary/script. The module also have access to **ResourceUsage**, which give the information about the resources on the agent/slave.

Commit:
compiling version with OpportunisticResourceEstimator

Commit:
remove unwanted file

Commit:
Merge branch 'master' of https://github.com/treefell/mesos-command-modules

Commit:
added ResourceEstimator to ModuleFactory

Commit:
entry point working

Commit:
added spawn process in OpportunisticResourceEstimator

Commit:
added mocked resources

Commit:
mocked resources working

Commit:
Oversubscribe doesn't crash

Commit:
Mesos agent running and commandRunner call the test script

Commit:
added output retrieving

Commit:
Convert Script.sh->return as Resources and set it as return of oversubscribable

Commit:
added link to input for the script.sh

Commit:
removing Json part and unused code

Commit:
move process spawn in initialize for ResourceUsage

Commit:
Resource usage envoyer au binaire

Commit:
change a log info to be easier to read

Commit:
remove comment

kamaradclimber commented on CMakeLists.txt:
```diff
@@ -60,6 +60,7 @@ set(MODULES_SOURCES
   ${CMAKE_SOURCE_DIR}/src/RunningContext.cpp
   ${CMAKE_SOURCE_DIR}/src/ConfigurationParser.cpp
   ${CMAKE_SOURCE_DIR}/src/ModulesFactory.cpp
+  ${CMAKE_SOURCE_DIR}/src/OpportunisticResourceEstimator.cpp
```
It should be named ResourceEstimator if nothing in its implementation is opportunistic

treefell replied:
Good point i have renamed it **CommandResourceEstimator** to avoid confusion with **mesos::ResourceEstimator**

kamaradclimber commented on src/ModulesFactory.cpp:
```diff
@@ -24,6 +25,12 @@ ::mesos::slave::Isolator* createIsolator(
+::mesos::slave::ResourceEstimator* createResourceEstimator(
+    const ::mesos::Parameters& parameters) {
+  Configuration cfg = ConfigurationParser::parse(parameters);
+  return new OpportunisticResourceEstimator(cfg.prepareCommand, cfg.isDebugSet);
```
why is it named prepareCommand? It should be something related to what the binary does (something like resourceEstimatorCommand)

treefell replied:
I forgot to change the Configuration file, i added a **oversubscribableCommand** variable since the function to use the command is **oversubscribable**

kamaradclimber commented on src/OpportunisticResourceEstimator.cpp:
```diff
@@ -0,0 +1,122 @@
+  Try<Resources> _resources = ::mesos::Resources::parse(
+      "[{\"name\" : \"cpus\", \"type\":\"SCALAR\", \"scalar\" : {\"value\" : "
+      "\"0\"}, \"role\" : \"*\"}]");
+  Resources resources = _resources.get();
```
do we need to return a valid resource if to set?
Shouldn't we return something like a Failure (the "null" case of Try class)

kamaradclimber replied:
@treefell could you answer my comment here?

treefell replied:
Actually try to return **NULL** and **nullptr** it does not compile while i do that (i get a problem of compatibility with resources). So i went for **resources** with _cpus_ at 0 as values. I will try to find another way.

kamaradclimber replied:
Try<Resources> means you can return a "success" with resources or a "failure" with an error message. Try returning a failure

kamaradclimber commented on src/OpportunisticResourceEstimator.cpp:
```diff
@@ -0,0 +1,122 @@
+  string input;
+  google::protobuf::TextFormat::PrintToString(usage, &input);
+
+  Try<string> output = cmdrunner.run(m_oversubscribableCommand.get(), input);
```
if the method _oversubscribable returns a future, it is probably better not to block if we can avoid it.
Given the way this method is used, maybe it shoudl not return a future but a Resource instead

treefell replied:
i have changed it so it return a **Resources**

kamaradclimber commented on src/OpportunisticResourceEstimator.cpp:
```diff
@@ -0,0 +1,122 @@
+
+  Try<string> output = cmdrunner.run(m_oversubscribableCommand.get(), input);
+  if (output.isError()) {
+    return resources;
```
I think we should:
- log the error
- return the error case of the Try class (unless mesos reacts weirdly to this)

treefell replied:
I have added a LOG(INFO) to retrieve the error
I am still investigating what i could give instead of resources

kamaradclimber replied:
ack

kamaradclimber reviewed (dismissed):
Note: I've reviewed only the beginning of the implementation. There are a few changes already.

This review needs to come with tests (even simple ones).

Commit:
rename Classes to CommandResourceEstimator instead of OpportunisticResourceEstimator

St0rmingBr4in commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,121 @@
+namespace criteo {
+namespace mesos {
+
+using namespace process;
+using namespace mesos;
```
`using namespace` is pure evil. This is a bad practice and you don't need it since you have the other `using` below :)

treefell replied:
hum i guess i was confused with the way namespace should be use. It have been removed

St0rmingBr4in commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,121 @@
+      const Option<Command>& oversubscribableCommand, bool isDebugMode);
+
+  virtual process::Future<Resources> oversubscribable();
+  virtual process::Future<Resources> _oversubscribable(
```
Even though this identifier is not causing any problems regarding the C++ standard, I would advise against starting an identifier with an underscore altogether.
```
Certain sets of names and function signatures are always reserved to the implementation:

Each name that contains a double underscore __ or begins with an underscore followed by an uppercase letter (2.12) is reserved to the implementation for any use.

Each name that begins with an underscore is reserved to the implementation for use as a name in the global namespace.
```
If you have trouble getting the right pointer to the overloaded function please look at https://en.cppreference.com/w/cpp/language/overloaded_address

treefell replied:
i have change most of the variable that had an _ in front of the name.

St0rmingBr4in commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,121 @@
+Future<Resources> CommandResourceEstimatorProcess::_oversubscribable(
+    const ResourceUsage& usage) {
+  // Mocking a resources to have valid return
+  Try<Resources> _resources = ::mesos::Resources::parse(
```
Same here, no underscore

St0rmingBr4in commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,121 @@
+    return resources;
+  }
+
+  Try<Resources> _cmdresources = ::mesos::Resources::parse(output.get());
```
Same

Commit:
rename opportunisticResourceEstimator.hpp to CommandResourceEstimator.hpp and remove some _ un naming

Lqp1 commented on src/CommandResourceEstimator.hpp:
```diff
@@ -0,0 +1,43 @@
+#ifndef __OPPORTUNISTIC_RESOURCE_ESTIMATOR_HPP__
+#define __OPPORTUNISTIC_RESOURCE_ESTIMATOR_HPP__
```
You renamed the file but not the guards ;)

treefell replied:
thanks for noticing it! should be fixed

Commit:
rename guard

kamaradclimber commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,120 @@
+  Try<string> output = cmdrunner.run(m_oversubscribableCommand.get(), input);
+  if (output.isError()) {
+    LOG(INFO) << output.error();
+    // return Failure("output from CommandRunner gave error: " +
```
Please don't commit commented code :)
In this case, it is indeed what I had in mind to return an error. Any reason not use it?

kamaradclimber replied:
ok let's keep the discussion in the first comment

kamaradclimber reviewed (dismissed):
Don't forget to add tests!

Commit:
segfault but should not

Commit:
added simple test for CommandResourceEstimator
Command Resource estimator
  -unexisting command
  -existing regular command
ModuleFactory for CommandResourceEstimator
  -empty command parameter
  -existing command parameter

Commit:
added modulefactory, and configurationParser for QoSController

Commit:
Added simple test for QoSController and modulefactory
Removed useless file

Commit:
added Module factory prototype

kamaradclimber commented on src/CommandQoSController.cpp:
```diff
@@ -0,0 +1,133 @@
+
+list<::mesos::slave::QoSCorrection> CommandQoSControllerProcess::m_corrections(
+    const ResourceUsage& m_usage) {
+  LOG(INFO) << "Start Corrections timer_start_module QoSController";
```
this would be more appropriate for debug level I think

treefell replied:
Yes this is more for debug, It is use to show time taken between different part of the module. like: start of module, start of the script, end of script, end of module

kamaradclimber commented on src/CommandQoSController.cpp:
```diff
@@ -0,0 +1,133 @@
+  // wished output = [{QoSCorrection}, {QoSCorrection}, ...]
+
+  std::string stroutput = output.get();
+  std::string delimiter = ";";
```
I'm not sure to get why the script could not return the correct format (which seems to be map nicely with json)

treefell replied:
Actually the problem come from the way message are supported.
Since the return must be an **array of QoSCorrection** there is noway to call for any message converting like **JsonStringtoMessage()**.
I tried to convert the **JsonArray to Array** in c++ but did not find and easy way to do it, a part from calling external library.

Each **QoSCorrection** look like this:
{"type":"KILL","kill":{"executorId":{"value":"app"},"frameworkId"{"value":"frameworkID"},"containerId":{"value":"containerID"}}}
Which kind of complicated the the way to split it so i use a special character as separator

If you have any good way to do i would gladly take it!

Lqp1 replied:
I agree with Greg, this part looks a bit fragile. I've read your answer but have somle additional questions:
* what is that "QoSscript timer_start_script QoSController" script? If this is something on our side, should it not return directly a json, like all other commands run?
* if it's not in our hands, maybe there should be an intermediate wrapper to do it anyway, because I don't feel this kind of things needs to be done in this module.

treefell replied:
thanks for the feeback, the "QoSscript timer_start_script QoScontroller" is an output line (just use to get a timestamp) to notify the start of the **commandRunner** running the **QoSscript.rb** that give the QoScorrection after receiving the ResourceUsage. 
It should totally give a JSON, i will try to rework it to return a JSON.

kamaradclimber commented on src/CommandQoSController.cpp:
```diff
@@ -0,0 +1,133 @@
+                                           const Option<Command>& _corrections,
+                                           bool isDebugMode)
+    : m_correctionsCommand(_corrections), m_isDebugMode(isDebugMode) {
+  LOG(INFO) << "new QoSController";
```
you can remove this line

treefell replied:
ack

kamaradclimber commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,126 @@
+    LOG(INFO) << output.error();
+    // return Failure("output from CommandRunner gave error: " +
+    // output.error());
+    return resources;
```
it sounds dangerous to return "0 cpus"  whenever the command fail. Maybe we could:
- store last result from the script (if any)
- otherwise, explicitely fail

treefell replied:
Actually i think returning **previous result** is more **dangerous** since if it keep failing it will always have the same amount thus **allocating more revocable task that the Agent** could take on.
Returning 0 prevent any revocable task to be allocated. In my tough if the Resource Estimator is broken revocable task should **not be able to allocate**.
I did try to fail, since the resource are returned to the slave it does not take fail from oversubscribable command, apart from modifying the slave i am not sure how we could make it work, if you have any suggestion i am interested

kamaradclimber commented on src/CommandResourceEstimator.cpp:
```diff
@@ -0,0 +1,126 @@
+    const std::string& name, const Option<Command>& _oversubscribable,
+    bool isDebugMode)
+    : m_oversubscribableCommand(_oversubscribable), m_isDebugMode(isDebugMode) {
+  LOG(INFO) << "new resourceEstimator";
```
you can drop this line

treefell replied:
ack

kamaradclimber reviewed (dismissed):
Thanks for the tests!

Commit:
remove LOG(INFO)

Commit:
remove LOG(INFO) for the timer timestamps

Commit:
added debug level for timer timestamps

Lqp1 commented on CMakeLists.txt:
```diff
@@ -72,6 +74,8 @@ set(MODULES_HEADERS
   ${CMAKE_SOURCE_DIR}/src/Helpers.hpp
   ${CMAKE_SOURCE_DIR}/src/Logger.hpp
   ${CMAKE_SOURCE_DIR}/src/ModulesFactory.hpp
+  ${CMAKE_SOURCE_DIR}/src/CommandResourceEstimator.hpp
```
Please provide a small documentation header in hpp files to quickly describe what are those methods supposed to do, what info they expect when running, ... You can get inspiration for the format if you look at Hook and Isolator headers.

treefell replied:
thanks for the feeback, i will make it more understandable!

Lqp1 commented on tests/scripts/oversubscribable_empty.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+# do nothing here, so output file remains empty
```
OPT: You can likely change your test to avoid the need of having an empty bash script, using some standard commands for example.

Lqp1 commented on tests/scripts/oversubscribable.sh:
```diff
@@ -0,0 +1,19 @@
+	 "role" : "*"}]
+EOM
+
+cat $INPUT_FILE > /tmp/test
```
You can remove this debug thing (and also comments above).

treefell replied:
ack

Commit:
remove comment and debug from tests/script/oversubscribable.sh
